### PR TITLE
Ghost tables visualization_backups migration STEP 1

### PR DESCRIPTION
**Pull request by oleurud**
Step 1 of the migration related to https://github.com/CartoDB/cartodb/issues/14698 

Remove VisualizationsExportService 

Not sure I am using the RoR way... don't be shy

### CODE REVIEW

File: db/migrate/20190313144818_revamp_visualization_backup.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
+require 'carto/db/migration_helper'
+
+include Carto::Db::MigrationHelper
+
+migration(
+  Proc.new do
+    run "TRUNCATE TABLE visualization_backups"
```
Comment (ethervoid) [inline db/migrate/20190313144818_revamp_visualization_backup.rb]: Not sure if we want to do this in a migration. I'd do it manually once we have the FF activated and nobody is using the service. What do you thing @javitonino

File: db/migrate/20190313144818_revamp_visualization_backup.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+    run "ALTER TABLE visualization_backups DROP CONSTRAINT visualization_backups_pkey"
+    drop_column :visualization_backups, :export_vizjson
+
+    add_column :visualization_backups, :id, :uuid, default: Sequel.lit('uuid_generate_v4()'), null: false
+    add_column :visualization_backups, :export, String, null: false
```
Comment (ethervoid) [inline db/migrate/20190313144818_revamp_visualization_backup.rb]: If we're going to store json, I'd use a json field or even a jsonb

File: db/migrate/20190313144818_revamp_visualization_backup.rb (hunk 1 above)
Comment (gonzaloriestra) [reply db/migrate/20190313144818_revamp_visualization_backup.rb]: I'd say truncate should be fast even for big tables, but I guess we cannot check it in staging because the table will be much smaller. Maybe we could use the task `purge_old_visualization_backups` first to avoid risks.

File: db/migrate/20190313144818_revamp_visualization_backup.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+include Carto::Db::MigrationHelper
+
+migration(
+  Proc.new do
+    run "TRUNCATE TABLE visualization_backups"
+
+    run "ALTER TABLE visualization_backups DROP CONSTRAINT visualization_backups_pkey"
+    drop_column :visualization_backups, :export_vizjson
+
+    add_column :visualization_backups, :id, :uuid, default: Sequel.lit('uuid_generate_v4()'), null: false
```
Comment (gonzaloriestra) [inline db/migrate/20190313144818_revamp_visualization_backup.rb]: Why can't you keep using the visualization_id as pk?

File: db/migrate/20190313144818_revamp_visualization_backup.rb (hunk 1 above)
Comment (ethervoid) [reply db/migrate/20190313144818_revamp_visualization_backup.rb]: The problem is to add that to a migration. If we rerun that migration for some reason we're going to wipe out all the data. Since this is a one time operation for sure, we should do it outside the migrations IMHO

Comment (gonzaloriestra) [reply db/migrate/20190313144818_revamp_visualization_backup.rb]: What about dropping and creating a new table then?

File: db/migrate/20190313144818_revamp_visualization_backup.rb (hunk 2 above)
Comment (oleurud) [reply db/migrate/20190313144818_revamp_visualization_backup.rb]: Yes, I forgot it!

File: db/migrate/20190313144818_revamp_visualization_backup.rb (hunk 3 above)
Comment (oleurud) [reply db/migrate/20190313144818_revamp_visualization_backup.rb]: Because we could have several visualization backups since we are going to save it when a visualization is destroyed (as usual) and also when a map layer is deleted.

File: db/migrate/20190313144818_revamp_visualization_backup.rb (hunk 1 above)
Comment (ethervoid) [reply db/migrate/20190313144818_revamp_visualization_backup.rb]: We have to rename it and create a new model I suppose but that table we have fits perfectly we only want to wipe out, as we did in the likes one, to start using it in other way

Comment (oleurud) [reply db/migrate/20190313144818_revamp_visualization_backup.rb]: If we rerun the task after some time running in production, the data will lose too, because the column with the export info will be dropped.

About dropping and creating a new table, the problem will be the same as the one warned by @ethervoid (unless we use different table names)

Comment (gonzaloriestra) [reply db/migrate/20190313144818_revamp_visualization_backup.rb]: @ethervoid I wouldn't say perfectly: we have to drop the PK, drop a column, add 2 columns and add a PK... 

@oleurud Yes, I meant to create a new table with a new name. That way we would avoid additional manual tasks.

Comment (oleurud) [reply db/migrate/20190313144818_revamp_visualization_backup.rb]: :thinking: If we don't change the column name, we could avoid the problem. I prefer to change it, but...

Comment (oleurud) [reply db/migrate/20190313144818_revamp_visualization_backup.rb]: @gonzaloriestra question as a big RoR padawan: if we change the table name, do we need to duplicate the `Visualization` class or do we have another option?

File: db/migrate/20190313144818_revamp_visualization_backup.rb (hunk 2 above)
Comment (oleurud) [reply db/migrate/20190313144818_revamp_visualization_backup.rb]: As we are not going to make queries usually, I am going to use `JSON`

File: db/migrate/20190313144818_revamp_visualization_backup.rb (hunk 1 above)
Comment (javitonino) [reply db/migrate/20190313144818_revamp_visualization_backup.rb]: If we want to do it the clean way, I would go for the nuclear option:
1. Remove all VisualizationsBackup code
2. Create a migration to DROP the existing table and create a new one. We can reuse the name.
3. Create a new model for VisualizationBackup with the new code

In any case we probably want to delete the existing model (using Sequel) and create a new one (using ActiveRecord). That way, it's easier to plug into the existing Visualization code (which uses ActiveRecord).

File: db/migrate/20190313144818_revamp_visualization_backup.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
+require 'carto/db/migration_helper'
+
+include Carto::Db::MigrationHelper
+
+migration(
+  Proc.new do
+    run "TRUNCATE TABLE visualization_backups"
+
+    run "ALTER TABLE visualization_backups DROP CONSTRAINT visualization_backups_pkey"
+    drop_column :visualization_backups, :export_vizjson
```
Comment (javitonino) [inline db/migrate/20190313144818_revamp_visualization_backup.rb]: This migration cannot be run without deleting all the existing models first, or at least adding code to tell the model that it shouldn't use this column even though it's in the table.

Reference: https://blog.codeship.com/rails-migrations-zero-downtime/

Comment (oleurud) [reply db/migrate/20190313144818_revamp_visualization_backup.rb]: Thanks for it :)

Comment (ethervoid) [reply db/migrate/20190313144818_revamp_visualization_backup.rb]: :classical_building:

File: app/services/carto/visualizations_export_service.rb [hunk 5]
```diff
@@ -28,22 +28,7 @@ def export(visualization_id)
...
-        backup_entry = Carto::VisualizationBackup.new(
-          username: visualization.user.username,
-          visualization: visualization.id,
-          export_vizjson: data
-        )
-        backup_entry.save
-
-        true
-      end
+      false
```
Comment (gonzaloriestra) [inline app/services/carto/visualizations_export_service.rb]: The previous lines of this method are useless now. But I guess you are you planning to reuse the code of this class later with the new table, right?

Comment (oleurud) [reply app/services/carto/visualizations_export_service.rb]: The first idea was to keep this PR as simple as possible, keeping some unused code.

But now, working on https://github.com/CartoDB/cartodb/pull/14745, I have found this way very time consuming (in addition to no sense), so right now I think I am going to apply the good way: remove the `visualizations_export_service` and related things.

If after removing all the stuff the tests are green, I will make the same here, because it is the right place to do it

cc/ @ethervoid

File: db/migrate/20190313144818_revamp_visualization_backup.rb [hunk 6]
```diff
@@ -0,0 +1,17 @@
+require 'carto/db/migration_helper'
+
+include Carto::Db::MigrationHelper
+
+migration(
+  Proc.new do
+    drop_table :visualization_backups
```
Comment (javitonino) [inline db/migrate/20190313144818_revamp_visualization_backup.rb]: You need to split this migration into a different PR. The code needs to be removed and deployed first. Then, you can do the migration to drop the table.

Comment (oleurud) [reply db/migrate/20190313144818_revamp_visualization_backup.rb]: Thank you again for the heads up.

For the record, in a deploy:
- start running the migration (with the previous code running in the server)
- then stop the server, pull the code and restart

So If I merge code and migration together, we will run the migration (removing a table) and the server will start to fail until the second step begins (it could take several minutes)
